enterprise_access/apps/api/v1/views/browse_and_request.py

```diff
@@ -797,6 +812,108 @@ def create(self, request, *args, **kwargs):
+        fn=get_enterprise_uuid_from_request_data,
+    )
+    @action(detail=False, url_path="decline", methods=["post"])
+    def decline(self, *args, **kwargs):
+        """
+        Action of declining a Learner Credit Subsidy Request
+        """
+        # Extract and validate input parameters
+        enterprise_customer_uuid = get_enterprise_uuid_from_request_data(self.request)
+        learner_credit_request_uuid = self.request.data.get("subsidy_request_uuid")
+        send_notification = self.request.data.get("send_notification", False)
+        disassociate_from_org = self.request.data.get("disassociate_from_org", False)
+
+        # Validate required parameters
+        if not learner_credit_request_uuid:
+            return Response(
+                "You must provide a subsidy request UUID to be declined.", status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            validate_uuid(learner_credit_request_uuid)
+        except ParseError:
+            return Response(
+                f"Subsidy Request UUID provided ({learner_credit_request_uuid}) is not a valid UUID",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # Fetch and validate the request
+        try:
+            learner_credit_request = LearnerCreditRequest.objects.get(uuid=learner_credit_request_uuid)
+        except LearnerCreditRequest.DoesNotExist:
+            return Response(
+                f"Learner Credit Request with UUID {learner_credit_request_uuid} not found.",
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        # State validation with action logging
+        if learner_credit_request.state not in [SubsidyRequestStates.REQUESTED]:
+            LearnerCreditRequestActions.create_action(
+                learner_credit_request=learner_credit_request,
+                recent_action=get_action_choice(SubsidyRequestStates.DECLINED),
+                status=get_user_message_choice(SubsidyRequestStates.REQUESTED),
+                error_reason=get_error_reason_choice(LearnerCreditRequestActionErrorReasons.FAILED_DECLINE),
+                traceback=f"Cannot decline request in state: {learner_credit_request.state}",
+            )
+
+            return Response(
+                f"Learner Credit Request with UUID {learner_credit_request_uuid} cannot be declined. "
+                f"Current state: {learner_credit_request.state}",
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+
+        # Perform the decline operation
+        try:
+            with transaction.atomic():
+                learner_credit_request.decline(self.user)
+                LearnerCreditRequestActions.create_action(
+                    learner_credit_request=learner_credit_request,
+                    recent_action=get_action_choice(SubsidyRequestStates.DECLINED),
+                    status=get_user_message_choice(SubsidyRequestStates.DECLINED),
+                )
```

Comment:
We are calling this block multiple times across different branches in the code.

How about we create the action instance before entering the atomic block and then in case of an error simply update that same instance with the error info?